src/Approximations/iterative_refinement.jl

```diff
@@ -1,82 +1,145 @@
+"""
+struct PolygonalOverapproximation{N<:AbstractFloat}
+    S::LazySet
+    approx_list::Vector{LocalApproximation{N}}
+end
+
+# initialize a polygonal overapproximation with an empty list
+PolygonalOverapproximation{N}(S::LazySet) where {N<:AbstractFloat} = PolygonalOverapproximation(S::LazySet, LocalApproximation{N}[])
+
+"""
+    new_approx(S::LazySet, p1::Vector{N}, d1::Vector{N}, p2::Vector{N}, d2::Vector{N}) where {N<:AbstractFloat}
+
+### Fields
 
-- `p1`        -- first inner point
-- `d1`        -- first direction
-- `p2`        -- second inner point
-- `d2`        -- second direction
+- `S`          -- convex set
+- `p1`         -- first inner point
+- `d1`         -- first direction
+- `p2`         -- second inner point
+- `d2`         -- second direction
 
 ### Output
 
-A new `Approximation2D` instance.
+A local approximation of `S` in the given directions.
 """
-function Approximation2D(p1::Vector{N},
-                         d1::Vector{N},
-                         p2::Vector{N},
-                         d2::Vector{N}) where {N<:AbstractFloat}
-    ndir = [p2[2]-p1[2], p1[1]-p2[1]]
-    norm_ndir = norm(ndir)
-
-    if norm_ndir > TOL_DIR
-        ndir = ndir/norm_ndir
-        q = intersection(Line(d1, dot(d1, p1)), Line(d2, dot(d2, p2)))
-        Approximation2D(p1, d1, p2, d2, q, dot(ndir, q) - dot(ndir, p1), ndir,
-                        true)
+function new_approx(S::LazySet, p1::Vector{N}, d1::Vector{N}, p2::Vector{N}, d2::Vector{N}) where {N<:AbstractFloat}
+    if norm(p1-p2, 2) < TOL
+        # this approximation cannot be refined and we set q = p1 by convention
+        ap = LocalApproximation{N}(p1, d1, p2, d2, p1, false, zero(N))
     else
-        Approximation2D(p1, d1, p2, d2, p1, zero(N), ndir, false)
+        ndir = normalize([p2[2]-p1[2], p1[1]-p2[1]])
+        q = intersection(Line(d1, dot(d1, p1)), Line(d2, dot(d2, p2)))
+        approx_error = min(norm(q - σ(ndir, S)), dot(ndir, q - p1))
+        refinable = (approx_error > TOL) && !(norm(p1-q, 2) < TOL || norm(q-p2, 2) < TOL)
+        ap = LocalApproximation{N}(p1, d1, p2, d2, q, refinable, approx_error)
     end
+    return ap
 end
 
 """
-    refine(S::LazySet, approx::Approximation2D)::Tuple{Approximation2D, Approximation2D}
+    addapproximation!(Ω::PolygonalOverapproximation, p1::Vector{N}, d1::Vector{N}, p2::Vector{N}, d2::Vector{N}) where {N <: AbstractFloat}
 
-Refine the given approximation.
+### Fields
+
+- `Ω`          -- polygonal overapproximation of a convex set
+- `p1`         -- first inner point
+- `d1`         -- first direction
+- `p2`         -- second inner point
+- `d2`         -- second direction
+
+### Output
+
+The list of local approximations in `Ω` of the set `Ω.S` is updated in-place and
+the new approximation is returned by this function.
+"""
+function addapproximation!(Ω::PolygonalOverapproximation, p1::Vector{N}, d1::Vector{N}, p2::Vector{N}, d2::Vector{N}) where {N <: AbstractFloat}
```

Comment:
long line
maybe add ``::Void`` for output type

Reply:
actually i was suggesting to return the computed approximation (see the Output) but the code was wrong so i've fixed it and the function's signature